### PR TITLE
Warn about unchecked type tests in primitive catch cases

### LINKED ISSUE #19013: `catch case` fails to warn in case of unchecked bounded type parameter
## Compiler version

3.3.1

## Minimized code

```Scala
def handle[E <: Exception](f: => Unit): Option[E] = 
  try 
    f
    None
  catch case e: E => Some(e)

val r: RuntimeException = handle[RuntimeException](throw new Exception()).get
```

## Output

```scala
java.lang.ClassCastException: class java.lang.Exception cannot be cast to class java.lang.RuntimeException (java.lang.Exception and java.lang.RuntimeException are in module java.base of loader 'bootstrap')
```

## Expectation

Fail compilation because type of E is not known in compile time, allow compilation for inline method (or maybe also by ClassTag in scope?).

### PR DESCRIPTION
**Pull request by hamzaremmal**
Closes #19013

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala [hunk 1]
```diff
@@ -381,6 +379,19 @@ object TypeTestsCasts {
     interceptWith(expr)
   }
 
+  /** After PatternMatcher, the only Bind nodes are present in simple try-catch trees
```
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala:382]: One extra "the"?

File: tests/neg/i4812.scala
Comment (hamzaremmal) [inline tests/neg/i4812.scala:1]: I think this file should be moved to `tests/warn`

File: tests/neg/i4812.check [hunk 2]
```diff
@@ -34,6 +34,12 @@
    |           the type test for A cannot be checked at runtime because it's a local class
    |
    | longer explanation available when compiling with `-explain`
+-- [E092] Pattern Match Error: tests/neg/i4812.scala:80:11 -------------------------------------------------------------
+80 |      case prev: A => prev.elem // error: the type test for A cannot be checked at runtime
+   |           ^
+   |           the type test for A cannot be checked at runtime because it's a local class
```
Comment (dwijnand) [inline tests/neg/i4812.check:40]: This is a sign you've overreached: you've hit a binding, but it lives inside an unchecked match, so you shouldn't emit the warning.

One way to fix this is doing something like https://github.com/lampepfl/dotty/pull/19190/files, where you record that you're in an unchecked match in the context, so you can honour it before emitting the warning.

File: tests/neg/i4812.scala
Comment (hamzaremmal) [reply tests/neg/i4812.scala:1]: So this is being done by another PR actually. See #18829

File: tests/neg/i4812.check (hunk 2 above)
Comment (hamzaremmal) [reply tests/neg/i4812.check]: That's actually simpler than that, we did a mistake when refactoring the code of `interceptTypeApply` by switching the parameters :-(

**Review by dwijnand — APPROVED** [review]

**Pull request merged**